File: serialportoak/src/main/java/com/open/serial/liboak/JniSerialPort.java
package com.open.serial.liboak;


import android.util.Log;

import java.io.FileDescriptor;

public class JniSerialPort {
    static {
        System.loadLibrary("serialportoak");
    }

    private static final char[] HEX = "0123456789ABCDEF".toCharArray();


    public static String bufferHexString(byte[] buffer, int off, int len) {
        StringBuilder sb = new StringBuilder(len * 2);
        int v;
        for (int i = 0; i < len; i++) {
            v = buffer[i + off];
            sb.append(HEX[v >> 4 & 0xF]).append(HEX[v & 0xF]);
        }
        return sb.toString();
    }


    public enum Parity {
        NONE, // 无校验
        ODD,  // 奇校验
        EVEN, // 偶校验
        MARK, // 标志位校验
        RTSCTS_IN, SPACE // 空位校验
    }

    public enum FlowControl {
        NONE,           // 无流控制
        RTSCTS_IN,          // RTS/CTS硬件流控制（输入）
        RTSCTS_OUT,         // RTS/CTS硬件流控制（输出）
        DSRDTR_IN,          // DSR/DTR硬件流控制（输入）
        DSRDTR_OUT          // DSR/DTR硬件流控制（输出）
    }

    /**
     * 开启串口
     *
     * @param path        串口设备路径  示例: "/dev/ttyS0"
     * @param baud        波特率       示例: 115200
     * @param dataBits    数据位       示例: 8
     * @param stopBits    停止位       示例: 1 2
     * @param parity      校验方式
     * @param flowControl 流控
     */
    public FileDescriptor open(
            String path,
            int baud,
            int dataBits,
            int stopBits,
            Parity parity,
            FlowControl flowControl
    ) {

        int parityInt;
        switch (parity) {
            case ODD:
                parityInt = 1;
                break;
            case EVEN:
                parityInt = 2;
                break;
            case MARK:
                parityInt = 3;
                break;
            case SPACE:
                parityInt = 4;
                break;
            default:
                parityInt = 0;
                break;
        }

        int flowControlInt;
        switch (flowControl) {
            case RTSCTS_IN:
                flowControlInt = 1;
                break;
            case RTSCTS_OUT:
                flowControlInt = 2;
                break;
            case DSRDTR_IN:
                flowControlInt = 3;
                break;
            case DSRDTR_OUT:
                flowControlInt = 4;
                break;
            default:
                flowControlInt = 0;
                break;
        }
        return open(path, baud, dataBits, stopBits, parityInt, flowControlInt);
    }

    private native FileDescriptor open(
            String path,
            int baud,
            int dataBits,
            int stopBits,
            int parity,
            int flowControl
    );

    /**
     * 关闭串口
     */
    public native int close();

    private JniSerialPort() {

    }

    private static final class H {
        private static final JniSerialPort i = new JniSerialPort();
    }

    /**
     * 单例可选
     */
    public static JniSerialPort getInstance() {
        return H.i;
    }
}

File: serialport/src/main/java/com/open/serial/lib/JniSerialPort.kt
package com.open.serial.lib

import java.io.FileDescriptor

class JniSerialPort {

    enum class Parity {
        NONE, // 无校验
        ODD,  // 奇校验
        EVEN, // 偶校验
        MARK, // 标志位校验
        SPACE // 空位校验
    }

    enum class FlowControl {
        NONE,               // 无流控制
        RTSCTS_IN,          // RTS/CTS硬件流控制（输入）
        RTSCTS_OUT,         // RTS/CTS硬件流控制（输出）
        DSRDTR_IN,          // DSR/DTR硬件流控制（输入）
        DSRDTR_OUT          // DSR/DTR硬件流控制（输出）
    }

    /**
     * 开启串口
     * @param path        串口设备路径  示例: "/dev/ttyS0"
     * @param baud        波特率       示例: 115200
     * @param dataBits    数据位       示例: 8
     * @param stopBits    停止位       示例: 0
     * @param parity      校验方式
     * @param flowControl 流控
     * */
    fun open(
        path: String,
        baud: Int,
        dataBits: Int = 8,
        stopBits: Int = 0,
        parity: Parity = Parity.NONE,
        flowControl: FlowControl = FlowControl.NONE
    ): FileDescriptor? {
        val parityInt = when (parity) {
            Parity.NONE -> 0
            Parity.ODD -> 1
            Parity.EVEN -> 2
            Parity.MARK -> 3
            Parity.SPACE -> 4
        }
        val flowControlInt = when (flowControl) {
            FlowControl.NONE -> 0
            FlowControl.RTSCTS_IN -> 1
            FlowControl.RTSCTS_OUT -> 2
            FlowControl.DSRDTR_IN -> 3
            FlowControl.DSRDTR_OUT -> 4
        }
        return open(path, baud, dataBits, stopBits, parityInt, flowControlInt)
    }

    private external fun open(
        path: String,
        baud: Int,
        dataBits: Int = 8,
        stopBits: Int = 0,
        parity: Int = 0,
        flowControl: Int = 0
    ): FileDescriptor?

    /**
     * 关闭串口
     * */
    external fun close(): Int


    companion object {

        private val HEX = "0123456789ABCDEF".toCharArray()

        init {
            System.loadLibrary("serialport")
        }

        fun bufferHexString(buffer: ByteArray, off: Int = 0, len: Int = buffer.size): String {
            val sb = StringBuilder(len * 2)
            for (i in 0..off + len) {
                (buffer[i].toInt() and 0xFF).let { b ->
                    sb.append(HEX[b ushr 4]).append(HEX[b and 0xF])
                }
            }
            return sb.toString()
        }
    }

}